### ISSUE TITLE
error: Previous Tip block tip height differs by 7. Please delete and resync (-D)

### ISSUE THREAD

**Issue by lekanovic** [bitpay/insight-api#311]
I keep on getting this error after a while.

Info:
* bitcoind 0.10.0
* insight-api branch master: e1341150d7ac457485d47cb78184412dab88597c

I have tried this 
* INSIGHT_FORCE_RPC_SYNC = 1
* node utils/sync.js -D -v --rpc
* I have removed rm ~/.insight directory
* in config.js I have commented out "//dataDir: dataDir,"

Summary
If I remove the ~/.insight directory and do the resync it works for a while but after a while I get this error. Since I am on testnet this is not a big problem since syncing testnet takes 45min. But for mainnet this is not feasible since it takes days to sync that one.

Any suggestion how to fix this problem, does anyone else have this problem. I all started out with the bitcoind version 0.10.0

**Comment by AugustoL** [comment]
Same problem here, I have to resync with testnet, but like @lekanovic  says, if it happens with livenet im in big trouble :)

**Comment by lekanovic** [comment]
Yes it happens quite often for me. I am surprised the more people have not seen this problem. This problem started for me when I switched to bitcoind v0.10.0

**Comment by M4v3R** [comment]
I just got the same problem. Anyone has a solution?

**Comment by matiu** [comment]
We had the same problem in testnet.

You will need to to insight; delete de database, and resync from scratch.

On Tue, Jun 9, 2015 at 2:14 PM, Maciej Trębacz <notifications@github.com>
wrote:

> I just got the same problem. Anyone has a solution?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitpay/insight-api/issues/311#issuecomment-110437293>.
>



-- 
BitPay.com

**Comment by AugustoL** [comment]
Same for me, delete db and resync :S, lets hope that this be caused only due to our upgrade to 0.1

**Comment by M4v3R** [comment]
@matiu Is it something that will happen more commonly and will it happen on main net? Deleting and resyncing DB is a pretty dirty workaround.

@AugustoL I did not update my client in the mean time and still got this error.

**Comment by AugustoL** [comment]
@M4v3R ho ok, well this happened to me only once, but I change insight for request to blockr.io after this, If this happen to me in production im fucked :)

**Comment by matiu** [comment]
On Tue, Jun 9, 2015 at 2:18 PM, Maciej Trębacz <notifications@github.com>
wrote:

> @matiu <https://github.com/matiu> Is it something that will happen more
> commonly and will it happen on main net? Deleting and resyncing DB is a
> pretty dirty workaround.
>

I think it is related to deep reorg, and dozends of blocks arriving at the
same time at testnet. I did not see  that error never in livenet.


> @AugustoL <https://github.com/AugustoL> I did not update my client in the
> mean time and still got this error.
>

We are running 0.9.5 ( and we recommend 0.9.5 for insight, since you can
user file syncing).


> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bitpay/insight-api/issues/311#issuecomment-110438199>.
>



-- 
Matías Alejo Garcia
@ematiu
Roads? Where we're going, we don't need roads!

**Comment by AugustoL** [comment]
@matiu ok, I will try with 0.9.5 again and see how it goes.

**Comment by posita** [comment]
> I think it is related to deep reorg, and dozends of blocks arriving at the same time at testnet. I did not see  that error never in livenet.

This is likely a matter of luck. You'd likely see the same thing with mainnet if something like [this](https://bitcoinmagazine.com/3668/bitcoin-network-shaken-by-blockchain-fork/) ever happened again. The fact that insight can't handle a reorg is a big hole. The consensus can change at any time. Only with mainnet, a rebuild can be painful, especially if all of one's nodes go offline around the same time (and they probably would, unless you had some on each side of the fork, which you can't guarantee).

> We are running 0.9.5 ( and we recommend 0.9.5 for insight, since you can user file syncing).

FYI, this work-around has an expiration date. Currently that date is unknown, but it's correlated to (e.g.) the next feature-related hard fork.

**Comment by posita** [comment]
BTW, I've seen this happen twice in the past three days on my testnet instances:

> `Previous Tip block height differs by [e.g., 101]. Please delete and resync (-D)`

**Comment by pnagurny** [comment]
We are developing a replacement for insight-api called [Bitcore Node](https://github.com/bitpay/bitcore-node) which does handle reorgs correctly. It will expose a compatible Insight API.

**Comment by posita** [comment]
@pnagurny, glad to hear it's on the radar for Bitcore Node. :+1: Will Bitcore Node expose a REST-ful API (like `insight-api` does)? Right now Bitcore Node's API looks like it's in-Node.js-only (unless I've missed something)?

**Comment by lekanovic** [comment]
@pnagurny That is great news! Would you recommend to try Bitcoin Node at this stage or is it still under development?

**Comment by pnagurny** [comment]
@posita At present, Bitcore Node has an RPC style API over socket.io, but I am currently working on an Insight compatible REST API. Later, we will come out with a new REST API which will be more consistent and an all around improvement over insight-api.

@lekanovic It's still in development, but you are welcome to try it out in a non-production environment!

**Comment by posita** [comment]
@lekanovic, I'll defer to @pnagurny for the official line, but from what I've been able to gather from [Gitter](https://gitter.im/bitpay/bitcore), Insight API is going away. The new world order will be something like:

```
+---------------------------+
| web browser               |
+------------------v--------+
                   |
+----------+-------v------+-+
|          | insight (UI) | |
| Node.js  +-------v------+ |
|          | Bitcore Node | |
+----------+-------v------+-+
| Bitcoin Core (bitcoind)   |
+---------------------------+
```

Which means if you want to use Bitcore Node in a SOA, you're going to have to write your own replacement for a REST-ful API like Insight API inside Node.js:

```
+---------------------------+
| REST client               |
+------------------v--------+
                   |
+----------+-------v------+-+
|          | <<your API>> <-+---- you're going to have to implement this piece yourself
| Node.js  +-------v------+ |
|          | Bitcore Node | |
+----------+-------v------+-+
| Bitcoin Core (bitcoind)   |
+---------------------------+
```

Where `<<your API>>` is your own, from-scratch JavaScript replacement for Insight API that lives inside Node.js and calls Bitcore Node code to get the required blockchain data. Or maybe it could be something like this, if you wanted to share indexes with Bitpay's block explorer:

```
+------------------------+ +----------------+
| web browser            | | REST client    |
+-----------------v------+ +------v---------+
                  |               | 
+---------+-------v------+-+------v-------+-+
|         | insight (UI) | | <<your API>> | |
| Node.js +-------v------+-+------v-------+ |
|         | Bitcore Node                  | |
+---------+---------------v---------------+-+
| Bitcoin Core (bitcoind)                   |
+-------------------------------------------+
```

Compare this to today's world:

```
+-------------+ +-------------+
| web browser | | REST client |
+-------v-----+ +-----v-------+
        |             | 
+-------v------+------v-------+
| Insight (UI) > Insight API  |
+--------------+------v-------+
| Bitcoin Core (bitcoind)     |
+-----------------------------+

```

This is speculation though. Again, I'll defer to @pnagurny for the real details.

**Comment by posita** [comment]
*UPDATE: Oops! Thankfully, @pnagurny beat me to it while I was working on my ASCII art! :+1: My apologies! (My original (speculative, and late) comment is preserved below.)*

**Comment by pnagurny** [comment]
We are envisioning Bitcore Node as the main piece of software that you need to download for bitcoin blockchain infrastructure. We are developing a module system which will allow anyone to add additional functionality. Think of it as the Drupal or Wordpress of bitcoin.

Bitcore Node builds bitcoind and has bindings directly to it so you don't have to run an additional bitcoind process.

Bitcore Node will listen over http, and modules can add their own routes for exposing additional REST API functionality or web views. Modules can tap into "hooks" for different events: like when a new block is added or removed. A module can index additional data from the bitcoin blockchain which can then be queried via the routes it exposes over http.

I am writing two modules right now: an Insight API module and an Insight UI module. The Insight API module will expose the current Insight API over http://localhost:port/api/insight. The Insight UI module will serve the static content for running a blockexplorer at http://localhost:port/insight. It will by default query the local instance of BItcore Node running.

We will be adding additional modules later for doing things like storing Copay wallets.

Things are still in development and things can change pretty quickly. So right now if you want to experiment with it and gives us your feedback that would be great! It won't be ready for production for several months probably.

**Comment by posita** [comment]
@pnagurny, thanks for the clarification! :+1: My apologies for any confusion I caused. :blush:

**Comment by lekanovic** [comment]
Wow thanks for the very detailed explanitation guys. I think your vision for Bitcore Node is really awesome. I will download it and try it out. Where can i follow the development for Bitcore Node?

**Comment by braydonf** [comment]
Should no longer be an issue on v0.3.0 release and bitcore-node.

**Comment by lekanovic** [comment]
Great that you have fixed this problem on v3.0.0 branch I will try it out as soont as I have time. Is the bitcore-node already in beta stage?

**Comment by braydonf** [comment]
It's been released, and you should be able to run `npm install bitcore -g` and have insight up and running. Take a look at https://bitcore.io/guides/full-node
